src/cargo_communication/request_actor.rs

```diff
@@ -199,3 +174,533 @@ where
+        let (sender_to_main, receiver_from_actor) = unbounded::<Message>();
+        let (sender_to_actor, receiver_from_cargo) = unbounded::<CargoMessage>();
+
+        let mut mock_cargo_handle = MockCargoHandler::new();
+        // There is no robust way to return ExitStatus hence we return Error. In consequence the
+        // status code of response is 2(Error).
+        mock_cargo_handle
+            .expect_join()
+            .returning(|| Err(io::Error::from(io::ErrorKind::Other)));
+        mock_cargo_handle
+            .expect_receiver()
+            .return_const(receiver_from_cargo);
+        let req_actor: RequestActor<Compile, MockCargoHandler<CargoMessage>> =
+            default_req_actor(sender_to_main, mock_cargo_handle);
+
+        // The channel is closed so the actor finishes its execution.
+        drop(sender_to_actor);
+
+        req_actor.run();
+
+        let mut settings = Settings::clone_current();
+        settings.add_redaction(".params.eventTime", TIMESTAMP);
+        settings.bind(|| {
+            assert_json_snapshot!(receiver_from_actor.recv().unwrap(), @r###"
+            {
+              "method": "build/taskStart",
+              "params": {
+                "eventTime": "timestamp",
+                "taskId": {
+                  "id": "test_origin_id"
+                }
+              }
+            }
+            "###);
+        });
+
+        settings.add_redaction(".params.taskId.id", RANDOM_TASK_ID);
+        settings.bind(|| {
+            assert_json_snapshot!(receiver_from_actor.recv().unwrap(), @r###"
+            {
+              "method": "build/taskStart",
+              "params": {
+                "data": {
+                  "target": {
+                    "uri": ""
+                  }
+                },
+                "dataKind": "compile-task",
+                "eventTime": "timestamp",
+                "taskId": {
+                  "id": "random_task_id",
+                  "parents": [
+                    "test_origin_id"
+                  ]
+                }
+              }
+            }
+            "###);
+            assert_json_snapshot!(receiver_from_actor.recv().unwrap(), @r###"
+            {
+              "method": "build/taskFinish",
+              "params": {
+                "eventTime": "timestamp",
+                "status": 2,
+                "taskId": {
+                  "id": "random_task_id"
+                }
+              }
+            }
+            "###);
+            assert_json_snapshot!(receiver_from_actor.recv().unwrap(), @r###"
+            {
+              "id": "test_req_id",
+              "result": {
+                "originId": "test_origin_id",
+                "statusCode": 2
+              }
+            }
+            "###);
+        });
+        no_more_msg(receiver_from_actor);
+    }
+
+    #[test]
+    fn cancel_with_cargo_handle() {
+        let (sender_to_main, receiver_from_actor) = unbounded::<Message>();
+
```

Comment:
Nie lepiej byłoby zwrócić w default_actor (actor, receiver_from_actor)? Bo to jest w każdym teście i sender_to_main, jedynie przekazywany do default_actor

Reply:
Poprawione